File: mobile/app/src/main/java/vn/lailua/app/ui/quiz/QuestionView.kt
package vn.lailua.app.ui.quiz

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Bookmark
import androidx.compose.material.icons.filled.BookmarkBorder
import androidx.compose.material.icons.filled.Replay
import androidx.compose.material.icons.automirrored.filled.VolumeOff
import androidx.compose.material.icons.automirrored.filled.VolumeUp
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import vn.lailua.app.LocalApp
import vn.lailua.app.data.Consequence
import vn.lailua.app.data.JunctionScene
import vn.lailua.app.logic.WhatIf
import vn.lailua.app.data.Question
import vn.lailua.app.data.RoadScene
import vn.lailua.app.ui.Card
import vn.lailua.app.ui.Chip
import vn.lailua.app.ui.scene.JunctionCanvas
import vn.lailua.app.ui.scene.JunctionPhase
import vn.lailua.app.ui.scene.RoadCanvas
import vn.lailua.app.ui.scene.RoadPhase
import vn.lailua.app.ui.theme.Asphalt

private val LETTERS = listOf("A", "B", "C", "D", "E", "F")

/**
 * Một câu hỏi: sa hình / tình huống → biển báo → đề bài → đáp án → giải thích.
 * @param revealed đã chấm (hiện đúng/sai + giải thích + mô phỏng thứ tự).
 */
@Composable
fun QuestionView(
    q: Question,
    index: Int,
    total: Int,
    selected: Int?,
    revealed: Boolean,
    bookmarked: Boolean,
    vehicle: String,
    onSelect: (Int) -> Unit,
    onBookmark: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val app = LocalApp.current
    val repo = app.repo
    val chapter = repo.chapter(q.chapter)
    // Tự đọc câu mới nếu người dùng bật trong mục Tôi; rời màn hình thì dừng đọc.
    val autoSpeak = app.store.state.value?.autoSpeak == true
    LaunchedEffect(q.id) { if (autoSpeak) app.speech.speak(vn.lailua.app.ui.Speech.questionText(q.text, q.options)) else app.speech.stop() }
    DisposableEffect(Unit) { onDispose { app.speech.stop() } }

    Column(modifier) {
        // Đáp án đang được mô phỏng: mặc định là đáp án người học chọn; bấm "Thử cách xử lý khác" để đổi.
        var sim by remember(q.id, revealed, selected) { mutableIntStateOf(selected ?: q.answer) }
        q.scene?.let { SceneStage(q, revealed, sim, vehicle) }

        if (q.signs.isNotEmpty()) {
            Row(Modifier.fillMaxWidth().padding(top = 12.dp), horizontalArrangement = Arrangement.spacedBy(12.dp, Alignment.CenterHorizontally)) {
                q.signs.forEachIndexed { i, code ->
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        repo.signBitmap(code)?.let { Image(it.asImageBitmap(), contentDescription = code, Modifier.size(84.dp)) }
                        if (q.signs.size > 1) Text("Biển ${i + 1}", color = Asphalt.muted, fontSize = 12.sp, fontWeight = FontWeight.Bold)
                    }
                }
            }
        }

        Row(Modifier.fillMaxWidth().padding(top = 14.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            Chip(if (total > 0) "Câu ${index + 1}/$total" else "Trạm ${index + 1}", color = Asphalt.lane.copy(alpha = 0.15f), fg = Asphalt.lane)
            Chip("${chapter.icon} ${chapter.short}")
            if (q.critical) Chip("⚠ ĐIỂM LIỆT", color = Asphalt.red.copy(alpha = 0.15f), fg = Color(0xFFFCA5A5))
            Spacer(Modifier.weight(1f))
            SpeakButton(q)
            IconButton(onClick = onBookmark, Modifier.size(32.dp)) {
                Icon(if (bookmarked) Icons.Filled.Bookmark else Icons.Filled.BookmarkBorder, contentDescription = "Lưu câu", tint = if (bookmarked) Asphalt.lane else Asphalt.muted)
            }
        }

        Text(q.text, color = Color.White, fontSize = 18.sp, fontWeight = FontWeight.Bold, lineHeight = 26.sp, modifier = Modifier.padding(top = 8.dp, bottom = 12.dp))

        q.options.forEachIndexed { i, opt ->
            val isAnswer = i == q.answer
            val isSel = i == selected
            val (bg, border, keyBg, keyFg) = when {
                revealed && isAnswer -> listOf(Asphalt.green.copy(alpha = 0.16f), Asphalt.green, Asphalt.green, Color.White)
                revealed && isSel -> listOf(Asphalt.red.copy(alpha = 0.16f), Asphalt.red, Asphalt.red, Color.White)
                isSel -> listOf(Asphalt.lane.copy(alpha = 0.12f), Asphalt.lane, Asphalt.lane, Color(0xFF0F172A))
                else -> listOf(Asphalt.surface, Asphalt.line, Color.White.copy(alpha = 0.08f), Color.White)
            }
            val shape = RoundedCornerShape(16.dp)
            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(bottom = 8.dp)
                    .clip(shape)
                    .background(bg)
                    .border(if (isSel || (revealed && isAnswer)) 2.dp else 1.dp, border, shape)
                    .clickable(enabled = !revealed) { onSelect(i) }
                    .padding(12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(Modifier.size(34.dp).clip(RoundedCornerShape(10.dp)).background(keyBg), contentAlignment = Alignment.Center) {
                    Text(LETTERS[i], color = keyFg, fontWeight = FontWeight.Black, fontSize = 15.sp)
                }
                Spacer(Modifier.width(12.dp))
                Text(opt, color = Color.White, fontSize = 16.sp, lineHeight = 22.sp, modifier = Modifier.weight(1f))
            }
        }

        AnimatedVisibility(revealed) {
            val ok = selected == q.answer
            Card(
                Modifier.fillMaxWidth().padding(top = 6.dp),
                color = if (ok) Asphalt.green.copy(alpha = 0.12f) else Asphalt.red.copy(alpha = 0.12f),
            ) {
                Column {
                    Text(
                        if (ok) "✅ Chính xác! Barie mở, đi tiếp thôi." else if (selected == null) "⏱ Chưa trả lời — đáp án đúng là ${LETTERS[q.answer]}." else "🚨 Sai rồi! Đáp án đúng là ${LETTERS[q.answer]}.",
                        color = if (ok) Color(0xFF6EE7B7) else Color(0xFFFCA5A5), fontWeight = FontWeight.ExtraBold, fontSize = 16.sp,
                    )
                    Text(q.explanation, color = Color.White.copy(alpha = 0.85f), fontSize = 15.sp, lineHeight = 22.sp, modifier = Modifier.padding(top = 8.dp))
                    q.tip?.let {
                        Row(Modifier.padding(top = 10.dp).clip(RoundedCornerShape(12.dp)).background(Asphalt.lane.copy(alpha = 0.12f)).padding(10.dp)) {
                            Text("💡 ", fontSize = 14.sp)
                            Text(it, color = Color(0xFFFDE68A), fontSize = 14.sp, lineHeight = 20.sp)
                        }
                    }
                }
            }
        }
        if (revealed && (q.consequences != null || q.scene is JunctionScene)) Consequences(q, selected, sim) { sim = it }
    }
}

/** Nút loa: đọc đề bài + đáp án bằng giọng nói. */
@Composable
private fun SpeakButton(q: Question) {
    val speech = LocalApp.current.speech
    val speaking by speech.speaking
    val available by speech.available
    if (!available) return
    IconButton(
        onClick = { if (speaking) speech.stop() else speech.speak(vn.lailua.app.ui.Speech.questionText(q.text, q.options)) },
        Modifier.size(32.dp),
    ) {
        Icon(
            if (speaking) Icons.AutoMirrored.Filled.VolumeOff else Icons.AutoMirrored.Filled.VolumeUp,
            contentDescription = if (speaking) "Dừng đọc" else "Đọc câu hỏi",
            tint = if (speaking) Asphalt.lane else Asphalt.muted,
        )
    }
}

private fun kindIcon(kind: String) = when (kind) { "crash" -> "💥"; "ticket" -> "🚓"; "danger" -> "⚠️"; else -> "✅" }
private fun kindTitle(kind: String) = when (kind) { "crash" -> "Va chạm"; "ticket" -> "Bị lập biên bản"; "danger" -> "Mất an toàn"; else -> "Đúng luật" }

/** Hậu quả khi chọn đáp án [choice]: sa hình dùng kịch bản WhatIf, cảnh lái dùng trường consequences. */
private fun outcomeOf(q: Question, choice: Int): Consequence? =
    WhatIf.plan(q, choice)?.verdict ?: if (choice == q.answer) null else q.consequences?.getOrNull(choice)

/**
 * Hậu quả của đáp án đang mô phỏng (va chạm / biên bản / mất an toàn) và bảng "Thử cách xử lý khác":
 * bấm một đáp án để xem điều gì xảy ra (sa hình diễn lại theo đáp án đó).
 */
@Composable
private fun Consequences(q: Question, selected: Int?, sim: Int, onSim: (Int) -> Unit) {
    val current = if (sim == q.answer) null else outcomeOf(q, sim)
    var open by remember(q.id) { mutableStateOf(false) }
    Column(Modifier.fillMaxWidth().padding(top = 10.dp)) {
        if (current != null) {
            val tone = if (current.kind == "crash") Asphalt.red else if (current.kind == "ok") Asphalt.green else Color(0xFFF59E0B)
            Row(
                Modifier.fillMaxWidth().clip(RoundedCornerShape(16.dp)).background(tone.copy(alpha = 0.14f)).border(1.dp, tone.copy(alpha = 0.5f), RoundedCornerShape(16.dp)).padding(12.dp),
                verticalAlignment = Alignment.Top,
            ) {
                Text(kindIcon(current.kind), fontSize = 26.sp)
                Spacer(Modifier.width(10.dp))
                Column {
                    Text(
                        (if (sim == selected) "Nếu làm vậy ngoài đường" else "Nếu chọn ${LETTERS[sim]}") + ": ${kindTitle(current.kind)}",
                        color = Color.White, fontWeight = FontWeight.ExtraBold, fontSize = 15.sp,
                    )
                    Text(current.text, color = Color.White.copy(alpha = 0.85f), fontSize = 14.sp, lineHeight = 20.sp, modifier = Modifier.padding(top = 2.dp))
                }
            }
        }
        Text(
            if (open) "▾ Thử cách xử lý khác" else "▸ Thử cách xử lý khác — xem hậu quả từng đáp án",
            color = Asphalt.lane, fontWeight = FontWeight.Bold, fontSize = 14.sp,
            modifier = Modifier.padding(top = 8.dp).clip(RoundedCornerShape(8.dp)).clickable { open = !open }.padding(vertical = 6.dp),
        )
        AnimatedVisibility(open) {
            Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                if (q.scene is JunctionScene) Text("Bấm một đáp án để xem sa hình diễn lại theo cách đó.", color = Asphalt.faint, fontSize = 12.sp)
                q.options.forEachIndexed { i, _ ->
                    val c = if (i == q.answer) null else outcomeOf(q, i)
                    val active = i == sim
                    Row(
                        Modifier.fillMaxWidth().clip(RoundedCornerShape(12.dp)).background(if (active) Asphalt.lane.copy(alpha = 0.12f) else Asphalt.surface)
                            .border(1.dp, if (active) Asphalt.lane.copy(alpha = 0.6f) else Color.Transparent, RoundedCornerShape(12.dp))
                            .clickable { onSim(i) }.padding(10.dp),
                        verticalAlignment = Alignment.Top,
                    ) {
                        Text(LETTERS[i], color = if (i == q.answer) Color(0xFF6EE7B7) else Color.White, fontWeight = FontWeight.Black, fontSize = 15.sp, modifier = Modifier.width(22.dp))
                        Text(if (c == null) "✅ " else "${kindIcon(c.kind)} ", fontSize = 14.sp)
                        Text(c?.text ?: "Xử lý đúng luật — đi tiếp an toàn.", color = Color.White.copy(alpha = 0.85f), fontSize = 14.sp, lineHeight = 20.sp)
                    }
                }
            }
        }
    }
}

/** Khung sa hình / tình huống với điều khiển pha hoạt hình. */
@Composable
private fun SceneStage(q: Question, revealed: Boolean, sim: Int, vehicle: String) {
    val correct = sim == q.answer
    val shape = RoundedCornerShape(20.dp)
    Box(Modifier.fillMaxWidth().height(230.dp).clip(shape).background(Color(0xFF1A2230)).border(1.dp, Asphalt.line, shape)) {
        when (val sc = q.scene) {
            is JunctionScene -> {
                var phase by remember(q.id) { mutableStateOf(JunctionPhase.INTRO) }
                var run by remember(q.id) { mutableIntStateOf(0) }
                val plan = remember(q.id, sim) { WhatIf.plan(q, sim) }
                LaunchedEffect(revealed, q.id, sim) { if (revealed) { run++; phase = JunctionPhase.PLAY } }
                JunctionCanvas(
                    sc, phase, runKey = run, plan = if (revealed) plan else null,
                    onIntroDone = { if (phase == JunctionPhase.INTRO) phase = JunctionPhase.IDLE },
                    onDone = { phase = JunctionPhase.DONE },
                    modifier = Modifier.fillMaxWidth().height(230.dp),
                )
                if (revealed && phase == JunctionPhase.DONE) {
                    IconButton(onClick = { run++; phase = JunctionPhase.PLAY }, Modifier.align(Alignment.TopEnd).padding(6.dp).clip(RoundedCornerShape(50)).background(Color(0xCC0F172A))) {
                        Icon(Icons.Filled.Replay, contentDescription = "Xem lại mô phỏng", tint = Color.White)
                    }
                }
            }
            is RoadScene -> {
                val phase = when {
                    !revealed -> RoadPhase.INTRO
                    correct -> RoadPhase.PASS
                    else -> RoadPhase.FAIL
                }
                RoadCanvas(sc.props, vehicle, phase, runKey = "${q.id}-$sim", seed = q.id, modifier = Modifier.fillMaxWidth().height(230.dp))
            }
            else -> {}
        }
    }
}

File: ios/LaiLua/UI/Quiz/QuestionView.swift
import SwiftUI
import LaiLuaCore

let LETTERS = ["A", "B", "C", "D", "E", "F"]

/**
 * Một câu hỏi: sa hình / tình huống → biển báo → đề bài → đáp án → giải thích.
 * - revealed: đã chấm (hiện đúng/sai + giải thích + mô phỏng thứ tự).
 */
struct QuestionView: View {
    let q: Question
    let index: Int
    let total: Int
    let selected: Int?
    let revealed: Bool
    let bookmarked: Bool
    let vehicle: String
    let onSelect: (Int) -> Void
    let onBookmark: () -> Void

    @EnvironmentObject private var app: AppContainer
    @EnvironmentObject private var store: ProgressStore
    /// Đáp án đang được mô phỏng khi người học bấm "Thử cách xử lý khác" (nil = đáp án đã chọn).
    @State private var simChoice: Int?

    private var sim: Int { simChoice ?? selected ?? q.answer }

    var body: some View {
        let chapter = app.repo.chapter(q.chapter)
        VStack(alignment: .leading, spacing: 0) {
            if q.scene != nil {
                SceneStage(q: q, revealed: revealed, sim: sim, vehicle: vehicle)
            }

            if !q.signs.isEmpty {
                HStack(alignment: .top, spacing: 12) {
                    ForEach(Array(q.signs.enumerated()), id: \.offset) { i, code in
                        VStack(spacing: 2) {
                            if let img = app.signs.image(code) {
                                Image(uiImage: img).resizable().frame(width: 84, height: 84).accessibilityLabel(code)
                            }
                            if q.signs.count > 1 { Text("Biển \(i + 1)").afont(12, .bold).foregroundColor(Asphalt.muted) }
                        }
                    }
                }
                .frame(maxWidth: .infinity)
                .padding(.top, 12)
            }

            HStack(spacing: 6) {
                Chip(total > 0 ? "Câu \(index + 1)/\(total)" : "Trạm \(index + 1)", color: Asphalt.lane.opacity(0.15), fg: Asphalt.lane)
                Chip("\(chapter?.icon ?? "") \(chapter?.short ?? "")")
                if q.critical { Chip("⚠ ĐIỂM LIỆT", color: Asphalt.red.opacity(0.15), fg: Asphalt.rose) }
                Spacer()
                SpeakButton(q: q)
                Button(action: onBookmark) {
                    Image(systemName: bookmarked ? "bookmark.fill" : "bookmark")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(bookmarked ? Asphalt.lane : Asphalt.muted)
                        .frame(width: 32, height: 32)
                }
                .accessibilityLabel("Lưu câu")
            }
            .padding(.top, 14)

            Text(q.text).afont(18, .bold).foregroundColor(.white).lineSpacing(4).fixedSize(horizontal: false, vertical: true)
                .padding(.top, 8).padding(.bottom, 12)

            ForEach(Array(q.options.enumerated()), id: \.offset) { i, opt in
                let isAnswer = i == q.answer
                let isSel = i == selected
                let style = optionStyle(isAnswer: isAnswer, isSel: isSel)
                Button { onSelect(i) } label: {
                    HStack(spacing: 12) {
                        Text(i < LETTERS.count ? LETTERS[i] : "\(i + 1)")
                            .afont(15, .black).foregroundColor(style.keyFg)
                            .frame(width: 34, height: 34)
                            .background(style.keyBg)
                            .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
                        Text(opt).afont(16).foregroundColor(.white).lineSpacing(3).multilineTextAlignment(.leading).fixedSize(horizontal: false, vertical: true)
                        Spacer(minLength: 0)
                    }
                    .padding(12)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(style.bg)
                    .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
                    .overlay(RoundedRectangle(cornerRadius: 16, style: .continuous).stroke(style.border, lineWidth: (isSel || (revealed && isAnswer)) ? 2 : 1))
                }
                .buttonStyle(.plain)
                .disabled(revealed)
                .padding(.bottom, 8)
            }

            if revealed {
                let ok = selected == q.answer
                CardView(color: ok ? Asphalt.green.opacity(0.12) : Asphalt.red.opacity(0.12)) {
                    VStack(alignment: .leading, spacing: 0) {
                        Text(ok ? "✅ Chính xác! Barie mở, đi tiếp thôi." : (selected == nil ? "⏱ Chưa trả lời — đáp án đúng là \(LETTERS[q.answer])." : "🚨 Sai rồi! Đáp án đúng là \(LETTERS[q.answer])."))
                            .afont(16, .heavy).foregroundColor(ok ? Asphalt.mint : Asphalt.rose).fixedSize(horizontal: false, vertical: true)
                        Text(q.explanation).afont(15).foregroundColor(.white.opacity(0.85)).lineSpacing(3).fixedSize(horizontal: false, vertical: true).padding(.top, 8)
                        if let tip = q.tip {
                            HStack(alignment: .top, spacing: 0) {
                                Text("💡 ").afont(14)
                                Text(tip).afont(14).foregroundColor(Color(hex: 0xFDE68A)).lineSpacing(2).fixedSize(horizontal: false, vertical: true)
                            }
                            .padding(10)
                            .background(Asphalt.lane.opacity(0.12))
                            .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
                            .padding(.top, 10)
                        }
                    }
                }
                .padding(.top, 6)
                .transition(.opacity.combined(with: .move(edge: .top)))
            }

            if revealed && (q.consequences != nil || q.junction != nil) {
                ConsequencesView(q: q, selected: selected, sim: sim) { simChoice = $0 }
            }
        }
        .animation(.easeOut(duration: 0.25), value: revealed)
        .onChange(of: q.id) { _ in simChoice = nil; autoSpeak() }
        .onChange(of: revealed) { _ in simChoice = nil }
        .onAppear { autoSpeak() }
        .onDisappear { app.speech.stop() }
    }

    /// Tự đọc câu mới nếu người dùng bật trong mục Tôi.
    private func autoSpeak() {
        if store.state.autoSpeak { app.speech.speak(Speech.questionText(q.text, options: q.options)) } else { app.speech.stop() }
    }

    private struct OptionStyle { let bg: Color; let border: Color; let keyBg: Color; let keyFg: Color }

    private func optionStyle(isAnswer: Bool, isSel: Bool) -> OptionStyle {
        if revealed && isAnswer { return OptionStyle(bg: Asphalt.green.opacity(0.16), border: Asphalt.green, keyBg: Asphalt.green, keyFg: .white) }
        if revealed && isSel { return OptionStyle(bg: Asphalt.red.opacity(0.16), border: Asphalt.red, keyBg: Asphalt.red, keyFg: .white) }
        if isSel { return OptionStyle(bg: Asphalt.lane.opacity(0.12), border: Asphalt.lane, keyBg: Asphalt.lane, keyFg: Asphalt.ink) }
        return OptionStyle(bg: Asphalt.surface, border: Asphalt.line, keyBg: Color.white.opacity(0.08), keyFg: .white)
    }
}

/// Nút loa: đọc đề bài + đáp án bằng giọng nói.
private struct SpeakButton: View {
    let q: Question
    @EnvironmentObject private var app: AppContainer

    var body: some View {
        if app.speech.available {
            SpeakButtonInner(q: q, speech: app.speech)
        }
    }
}

private struct SpeakButtonInner: View {
    let q: Question
    @ObservedObject var speech: Speech

    var body: some View {
        Button {
            if speech.speaking { speech.stop() } else { speech.speak(Speech.questionText(q.text, options: q.options)) }
        } label: {
            Image(systemName: speech.speaking ? "speaker.slash.fill" : "speaker.wave.2.fill")
                .font(.system(size: 15, weight: .semibold))
                .foregroundColor(speech.speaking ? Asphalt.lane : Asphalt.muted)
                .frame(width: 32, height: 32)
        }
        .accessibilityLabel(speech.speaking ? "Dừng đọc" : "Đọc câu hỏi")
    }
}

private func kindIcon(_ kind: String) -> String {
    switch kind { case "crash": return "💥"; case "ticket": return "🚓"; case "danger": return "⚠️"; default: return "✅" }
}

private func kindTitle(_ kind: String) -> String {
    switch kind { case "crash": return "Va chạm"; case "ticket": return "Bị lập biên bản"; case "danger": return "Mất an toàn"; default: return "Đúng luật" }
}

/// Hậu quả của đáp án đang mô phỏng và bảng "Thử cách xử lý khác": bấm một đáp án để xem điều gì xảy ra
/// (sa hình diễn lại theo đáp án đó).
private struct ConsequencesView: View {
    let q: Question
    let selected: Int?
    let sim: Int
    let onSim: (Int) -> Void
    @State private var open = false

    var body: some View {
        let current = sim == q.answer ? nil : WhatIf.outcome(q, choice: sim)
        VStack(alignment: .leading, spacing: 6) {
            if let current {
                let tone = current.kind == "crash" ? Asphalt.red : (current.kind == "ok" ? Asphalt.green : Color(hex: 0xF59E0B))
                HStack(alignment: .top, spacing: 10) {
                    Text(kindIcon(current.kind)).font(.system(size: 26))
                    VStack(alignment: .leading, spacing: 2) {
                        Text("\(sim == selected ? "Nếu làm vậy ngoài đường" : "Nếu chọn \(LETTERS[sim])"): \(kindTitle(current.kind))")
                            .afont(15, .heavy).foregroundColor(.white).fixedSize(horizontal: false, vertical: true)
                        Text(current.text).afont(14).foregroundColor(.white.opacity(0.85)).lineSpacing(2).fixedSize(horizontal: false, vertical: true)
                    }
                    Spacer(minLength: 0)
                }
                .padding(12)
                .background(tone.opacity(0.14))
                .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
                .overlay(RoundedRectangle(cornerRadius: 16, style: .continuous).stroke(tone.opacity(0.5), lineWidth: 1))
            }
            Button { withAnimation { open.toggle() } } label: {
                Text(open ? "▾ Thử cách xử lý khác" : "▸ Thử cách xử lý khác — xem hậu quả từng đáp án")
                    .afont(14, .bold).foregroundColor(Asphalt.lane).padding(.vertical, 6)
            }
            .buttonStyle(.plain)
            if open {
                if q.junction != nil {
                    Text("Bấm một đáp án để xem sa hình diễn lại theo cách đó.").afont(12).foregroundColor(Asphalt.faint)
                }
                ForEach(q.options.indices, id: \.self) { i in
                    let c = i == q.answer ? nil : WhatIf.outcome(q, choice: i)
                    let active = i == sim
                    Button { onSim(i) } label: {
                        HStack(alignment: .top, spacing: 6) {
                            Text(LETTERS[i]).afont(15, .black).foregroundColor(i == q.answer ? Asphalt.mint : .white).frame(width: 22, alignment: .leading)
                            Text(c.map { kindIcon($0.kind) } ?? "✅").afont(14)
                            Text(c?.text ?? "Xử lý đúng luật — đi tiếp an toàn.").afont(14).foregroundColor(.white.opacity(0.85))
                                .multilineTextAlignment(.leading).fixedSize(horizontal: false, vertical: true)
                            Spacer(minLength: 0)
                        }
                        .padding(10)
                        .background(active ? Asphalt.lane.opacity(0.12) : Asphalt.surface)
                        .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
                        .overlay(RoundedRectangle(cornerRadius: 12, style: .continuous).stroke(active ? Asphalt.lane.opacity(0.6) : .clear, lineWidth: 1))
                    }
                    .buttonStyle(.plain)
                }
            }
        }
        .padding(.top, 10)
    }
}

/// Khung sa hình / tình huống với điều khiển pha hoạt hình.
private struct SceneStage: View {
    let q: Question
    let revealed: Bool
    /// Đáp án đang mô phỏng (sa hình diễn lại theo đáp án này).
    let sim: Int
    let vehicle: String

    @State private var phase: JunctionPhase = .intro
    @State private var run = 0

    var body: some View {
        let correct = sim == q.answer
        ZStack(alignment: .topTrailing) {
            switch q.scene {
            case .some(.junction(let sc)):
                JunctionCanvas(
                    spec: sc, phase: phase, runKey: run,
                    onIntroDone: { if phase == .intro { phase = .idle } },
                    onDone: { phase = .done },
                    plan: revealed ? WhatIf.plan(q, choice: sim) : nil
                )
                if revealed && phase == .done {
                    Button { run += 1; phase = .play } label: {
                        Image(systemName: "arrow.counterclockwise").font(.system(size: 16, weight: .bold)).foregroundColor(.white)
                            .frame(width: 36, height: 36).background(Color(hex: 0x0F172A, alpha: 0.8)).clipShape(Circle())
                    }
                    .padding(6)
                    .accessibilityLabel("Xem lại mô phỏng")
                }
            case .some(.road(let sc)):
                let rp: RoadPhase = !revealed ? .intro : (correct ? .pass : .fail)
                RoadCanvas(props: sc.props, vehicle: vehicle, phase: rp, runKey: q.id * 10 + sim, seed: q.id)
            case .none:
                EmptyView()
            }
        }
        .frame(maxWidth: .infinity)
        .frame(height: 230)
        .background(Color(hex: 0x1A2230))
        .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 20, style: .continuous).stroke(Asphalt.line, lineWidth: 1))
        .onAppear { if revealed { run += 1; phase = .play } }
        .onChange(of: revealed) { r in if r { run += 1; phase = .play } }
        .onChange(of: sim) { _ in if revealed { run += 1; phase = .play } }
    }
}
